library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -1223,19 +1223,29 @@ impl<T, A: Allocator> VecDeque<T, A> {
 
     /// Given a range into the logical buffer of the deque, this function
     /// return two ranges into the physical buffer that correspond to
-    /// the given range.
-    fn slice_ranges<R>(&self, range: R) -> (Range<usize>, Range<usize>)
+    /// the given range. The `len` parameter should usually just be `self.len`;
+    /// the reason it's passed explicitly is that if the deque is wrapped in
+    /// a `Drain`, then `self.len` is not actually the length of the deque.
+    ///
+    /// # Safety
+    ///
+    /// This function is always safe to call. For the resulting ranges to be valid
+    /// ranges into the physical buffer, the caller must ensure that for all possible
+    /// values of `range` and `len`, the result of calling `slice::range(range, ..len)`
+    /// represents a valid range into the logical buffer, and that all elements
```

Comment:
That is indeed a bit of a mouthful and requires one to look up with `slice::range` does (I wasn't familiar with it). Maybe it's better to say "up to range's end or len, whichever is lower"?

Reply:
Something like "all elements up to ..." wouldn't cover all our use cases, as `Drain::as_slices` calls this function on a deque whose elements in the logical range `drain.start..drain.idx` aren't initialized anymore.

Reply:
I suppose we could say "all elements in the logical range `range.start..cmp::min(range.end, len)` must be initialized, but I'm not sure if that makes it more readable than it currently is.

Reply:
Also, since there's quite a few types that implement `RangeBounds`, I opted for a more precise and perhaps more convoluted comment over one that's simpler but might not cover all use cases or might introduce some ambiguities.

Reply:
Ok, but even if we keep the `slice::range` phrasing I'm still having trouble with the whole paragraph.

> the caller must ensure that for all possible values of `range` and `len`, the result of calling `slice::range(range, ..len)` represents a valid range into the logical buffer

What does the "for all possible values" add? Does that refer to the covered elements? Or does it just say that every time you pass in garbage you get garbage out?

Reply:
Eh, the "for all possible values" can just be left out. I'm not sure why I put it there in the first place so I guess I'll just remove it.